pkg/controller/stackconfigpolicy/elasticsearch_config_settings.go

```diff
@@ -0,0 +1,149 @@
+	SecretsMountKey                                  = "secretMounts.json"
+	ElasticsearchConfigAndSecretMountsHashAnnotation = "policy.k8s.elastic.co/elasticsearch-config-mounts-hash" //nolint:gosec
+	SourceSecretAnnotationName                       = "policy.k8s.elastic.co/source-secret-name"               //nolint:gosec
+)
+
+func newElasticsearchConfigSecret(policy policyv1alpha1.StackConfigPolicy, es esv1.Elasticsearch) (corev1.Secret, error) {
+	secretMountBytes, err := json.Marshal(policy.Spec.Elasticsearch.SecretMounts)
+	if err != nil {
+		return corev1.Secret{}, err
+	}
+	elasticsearchAndMountsConfigHash := getElasticsearchConfigAndMountsHash(policy.Spec.Elasticsearch.Config, policy.Spec.Elasticsearch.SecretMounts)
+	var configDataJSONBytes []byte
+	if policy.Spec.Elasticsearch.Config != nil {
+		configDataJSONBytes, err = policy.Spec.Elasticsearch.Config.MarshalJSON()
+		if err != nil {
+			return corev1.Secret{}, err
+		}
+	}
+	elasticsearchConfigSecret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: es.Namespace,
+			Name:      esv1.StackConfigElasticsearchConfigSecretName(es.Name),
+			Labels: eslabel.NewLabels(types.NamespacedName{
+				Name:      es.Name,
+				Namespace: es.Namespace,
+			}),
+			Annotations: map[string]string{
+				ElasticsearchConfigAndSecretMountsHashAnnotation: elasticsearchAndMountsConfigHash,
+			},
+		},
+		Data: map[string][]byte{
+			ElasticSearchConfigKey: configDataJSONBytes,
+			SecretsMountKey:        secretMountBytes,
+		},
+	}
+
+	// Set Elasticsearch as the soft owner
+	filesettings.SetSoftOwner(&elasticsearchConfigSecret, policy)
+
+	// Add label to delete secret on deletion of the stack config policy
+	elasticsearchConfigSecret.Labels[commonlabels.StackConfigPolicyOnDeleteLabelName] = commonlabels.OrphanObjectDeleteOnPolicyDelete
+
+	return elasticsearchConfigSecret, nil
+}
+
+func reconcileSecret(ctx context.Context,
+	c k8s.Client,
+	expected corev1.Secret,
+	owner client.Object) error {
+	reconciled := &corev1.Secret{}
+	return reconciler.ReconcileResource(reconciler.Params{
+		Context:    ctx,
+		Client:     c,
+		Owner:      owner,
+		Expected:   &expected,
+		Reconciled: reconciled,
+		NeedsUpdate: func() bool {
+			return !maps.IsSubset(expected.Labels, reconciled.Labels) ||
+				!maps.IsSubset(expected.Annotations, reconciled.Annotations) ||
+				!reflect.DeepEqual(expected.Data, reconciled.Data)
+		},
+		UpdateReconciled: func() {
+			reconciled.Labels = maps.Merge(reconciled.Labels, expected.Labels)
+			reconciled.Annotations = maps.Merge(reconciled.Annotations, expected.Annotations)
+			reconciled.Data = expected.Data
+		},
+	})
+}
+
+// reconcileSecretMounts creates the secrets in SecretMounts to the respective Elasticsearch namespace where they should be mounted to.
+func reconcileSecretMounts(ctx context.Context, c k8s.Client, es esv1.Elasticsearch, policy *policyv1alpha1.StackConfigPolicy) error {
```

Comment:
I'm not sure to understand if `Secrets` which are no longer referenced in a policy are deleted.

Reply:
currently if a secret is removed from a stack config policy, that secret gets left behind(it is deleted only when the stack config policy itself is deleted or the elasticsearch cluster is). I was going to address this in a separate PR, as even today we don't reset a secret that is no longer referenced in the stackconfig policy I believe.